README.md
# NAME

snippy.psgi - minimalist note-taking web application

# VERSION

version 0.001

# USAGE

    plackup -r bin/snippy.psgi

# OPTIONS

None.

# EXIT STATUS

See [plackup](https://metacpan.org/pod/plackup) or your favorite PSGI runner.

# SUPPORT

## Perldoc

You can find documentation for this module with the perldoc command.

    perldoc bin::snippy.psgi

## Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

- Search CPAN

    The default CPAN search engine, useful to view POD in HTML format.

    [http://search.cpan.org/dist/WebApp-Snippy](http://search.cpan.org/dist/WebApp-Snippy)

- AnnoCPAN

    The AnnoCPAN is a website that allows community annotations of Perl module documentation.

    [http://annocpan.org/dist/WebApp-Snippy](http://annocpan.org/dist/WebApp-Snippy)

- CPAN Ratings

    The CPAN Ratings is a website that allows community ratings and reviews of Perl modules.

    [http://cpanratings.perl.org/d/WebApp-Snippy](http://cpanratings.perl.org/d/WebApp-Snippy)

- CPANTS

    The CPANTS is a website that analyzes the Kwalitee ( code metrics ) of a distribution.

    [http://cpants.cpanauthors.org/dist/WebApp-Snippy](http://cpants.cpanauthors.org/dist/WebApp-Snippy)

- CPAN Testers

    The CPAN Testers is a network of smokers who run automated tests on uploaded CPAN distributions.

    [http://www.cpantesters.org/distro/W/WebApp-Snippy](http://www.cpantesters.org/distro/W/WebApp-Snippy)

- CPAN Testers Matrix

    The CPAN Testers Matrix is a website that provides a visual overview of the test results for a distribution on various Perls/platforms.

    [http://matrix.cpantesters.org/?dist=WebApp-Snippy](http://matrix.cpantesters.org/?dist=WebApp-Snippy)

- CPAN Testers Dependencies

    The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

    [http://deps.cpantesters.org/?module=bin::snippy.psgi](http://deps.cpantesters.org/?module=bin::snippy.psgi)

## Bugs / Feature Requests

Please report any bugs or feature requests through the web
interface at [https://github.com/mjgardner/Snippy/issues](https://github.com/mjgardner/Snippy/issues). You will be automatically notified of any
progress on the request by the system.

## Source Code

The code is open to the world, and available for you to hack on. Please feel free to browse it and play
with it, or whatever. If you want to contribute patches, please send me a diff or prod me to pull
from your repository :)

[https://github.com/mjgardner/Snippy](https://github.com/mjgardner/Snippy)

    git clone git://github.com/mjgardner/Snippy.git

# AUTHOR

Mark Gardner <mjgardner@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Mark Gardner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for WebApp-Snippy

{{$NEXT}}
    - initial release


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

# Normalize version strings like 6.30_02 to 6.3002,
# so that we can do numerical comparisons on it.
my $eumm_version = $ExtUtils::MakeMaker::VERSION;
$eumm_version =~ s/_//;

WriteMakefile(
    NAME                => 'WebApp::Snippy',
    AUTHOR              => q{Mark Gardner <mjgardner@cpan.org>},
    VERSION_FROM        => 'lib/WebApp/Snippy.pm',
    ABSTRACT            => 'A simple notetaking web application',
    ($eumm_version >= 6.3001
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More'            => 0,
        'YAML'                  => 0,
        'Dancer2'               => 0.205000,
        'Dancer2::Plugin::DBIC' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'WebApp-Snippy-*' },
);


schema_conf.pl
package main;

## no critic (Lax::RequireEndWithTrueConst, Modules::RequireEndWithOne)

use strict;
use warnings;
use utf8;
our $VERSION = '0.001';

{   schema_class   => 'WebApp::Snippy::Schema',
    lib            => './lib',
    connect_info   => { dsn => 'dbi:SQLite:./snippy.db' },
    loader_options => {
        dump_directory        => './lib',
        filter_generated_code => 'perltidy --profile=xt/author/perltidy.rc',
        components            => [qw(InflateColumn::DateTime TimeStamp)],
        generate_pod          => 0,
        custom_column_info    => sub {
            my ( $table, $column_name, $column_info ) = @_;

            for ($column_name) {
                'created' eq $_ and return {
                    data_type     => 'datetime',
                    set_on_create => 1,
                };
                'updated' eq $_ and return {
                    data_type     => 'datetime',
                    set_on_create => 1,
                    set_on_update => 1,
                };
            }
        },
    },
}


lib/WebApp/Snippy.pm
package WebApp::Snippy;

# ABSTRACT: minimalist note-taking web application

=head1 DESCRIPTION

This web application was developed as a CAP499 Capstone Project for
L<Independence University|http://independence.edu/>. It demonstrates a
minimal CRUD (create, report, update, delete) interface to a database
table accessed via L<DBIx::Class|DBIx::Class>.

=head1 SYNOPSIS

    use WebApp::Snippy;
    WebApp::Snippy->to_app;

=cut

use strict;
use warnings;
use utf8;
use Dancer2;
use Dancer2::Plugin::DBIC;

# VERSION

get '/:id?' => sub {
    my $id      = route_parameters->get('id');
    my $query   = query_parameters->get('query');
    my $note_rs = resultset('Note');

    my @results
        = $id ? $note_rs->find( { id => $id } )
        : length $query ? $note_rs->search(
        {   -or => [
                note   => { like => "%$query%" },
                source => { like => "%$query%" },
            ],
        },
        )
        : $note_rs->all;
    redirect q{/} if not ref $results[0];

    template index => {
        ( id => $id ) x !!$id,
        query   => $query,
        results => [ map { ( { _result_fields($_) } ) } @results, ],
    };
};

post '/:id?' => sub {
    my $id      = route_parameters->get('id');
    my %post    = map { $_ => body_parameters->get($_) } qw(note source);
    my $note_rs = resultset('Note');

    if ($id) {
        my $note = $note_rs->find( { id => $id } );
        redirect q{/} if not $note;

        if ( body_parameters->get('delete') ) {
            $note->delete();
            redirect q{/};
        }
        else { $note->update( \%post ) }
    }
    else {
        my $result = $note_rs->create( \%post );
        $id = $result->get_column('id');
    }
    redirect "/$id";
};

## no critic (Subroutines::RequireArgUnpacking)
sub _result_fields {
    return map { $_ => $_[0]->$_ } qw(id note source created updated);
}

1;


lib/WebApp/Snippy/Schema.pm
use utf8;

package WebApp::Snippy::Schema;

# Created by DBIx::Class::Schema::Loader
# DO NOT MODIFY THE FIRST PART OF THIS FILE

use strict;
use warnings;

use base 'DBIx::Class::Schema';

__PACKAGE__->load_namespaces;


# Created by DBIx::Class::Schema::Loader v0.07047 @ 2017-09-14 12:28:49
# DO NOT MODIFY THIS OR ANYTHING ABOVE! md5sum:dVJJ2bN89e3A7BkhVhngUg

# You can replace this text with custom code or comments, and it will be preserved on regeneration
# ABSTRACT: Database schema for WebApp::Snippy
# VERSION

1;

__END__

=head1 SYNOPSIS

    use WebApp::Snippy::Schema;
    my $schema = WebApp::Snippy::Schema->connect('dbi:SQLite:snippy.db');
    my @all_notes = $schema->resultset('Notes')->all;

=head1 DESCRIPTION

This class enables access to the database schema used by
L<WebApp::Snippy|WebApp::Snippy>.



lib/WebApp/Snippy/Schema/Result/Note.pm
use utf8;

package WebApp::Snippy::Schema::Result::Note;

# Created by DBIx::Class::Schema::Loader
# DO NOT MODIFY THE FIRST PART OF THIS FILE

use strict;
use warnings;

use base 'DBIx::Class::Core';
__PACKAGE__->load_components( "InflateColumn::DateTime", "TimeStamp" );
__PACKAGE__->table("notes");
__PACKAGE__->add_columns(
    "id",
    { data_type => "integer", is_auto_increment => 1, is_nullable => 0 },
    "note",
    { data_type => "text", is_nullable => 1 },
    "source",
    { data_type => "text", is_nullable => 1 },
    "created",
    { data_type => "datetime", is_nullable => 1, set_on_create => 1 },
    "updated",
    {   data_type     => "datetime",
        is_nullable   => 1,
        set_on_create => 1,
        set_on_update => 1,
    },
);
__PACKAGE__->set_primary_key("id");


# Created by DBIx::Class::Schema::Loader v0.07047 @ 2017-09-14 12:28:49
# DO NOT MODIFY THIS OR ANYTHING ABOVE! md5sum:0OTance9Xg7JHdAlROUDFw

# You can replace this text with custom code or comments, and it will be preserved on regeneration
# ABSTRACT: Result set for notes stored in the database
# VERSION

1;

__END__

=head1 SYNOPSIS

    use WebApp::Snippy::Schema;
    my $schema = WebApp::Snippy::Schema->connect('dbi:SQLite:snippy.db');
    my @all_notes = $schema->resultset('Notes')->all;

=head1 DESCRIPTION

This class enables access to the C<notes> table in
L<WebApp::Snippy::Schema|WebApp::Snippy::Schema>.

=head1 COMPONENTS LOADED

=over 4

=item * L<DBIx::Class::InflateColumn::DateTime|DBIx::Class::InflateColumn::DateTime>

=item * L<DBIx::Class::TimeStamp|DBIx::Class::TimeStamp>

=back

=attr id

Auto-incrementing integer primary key for the table.

=attr note

Text of the note. Can be undef.

=attr source

Source attribution of the note, usually a URL. Can be undef.

=attr created

L<DateTime|DateTime> when the note was created. Can be undef. Set
automatically on record creation.

=attr updated

L<DateTime|DateTime> when the note was updated. Can be undef. Set
automatically on record creation and update.
